File: android/app/src/main/java/ai/oriveo/community/core/provider/CapabilityControlPresentation.kt
package ai.oriveo.community.core.provider

import ai.oriveo.community.core.data.remote.MetadataClient
import ai.oriveo.community.core.data.remote.canonicalCapabilityTransport
import ai.oriveo.community.core.model.AIModel
import ai.oriveo.community.core.model.GenerationProfileRef
import ai.oriveo.community.core.model.LocalCapabilityCustomFragmentStore
import ai.oriveo.community.core.model.Provider

/**
 * The one projection from a capability's control state onto what the UI is allowed to show.
 *
 * [CapabilityControlResolution.Verdict] only carries a handful of wire states, but the panel
 * has to answer finer questions than that: is "unsupported" because the capability does not
 * exist, or because it only lives inside a separate external connector? Is "unknown" a route
 * that has not been published yet (worth waiting for), or one that was never configured for
 * this model (switch models instead)? Collapsing those into a single `unknown` leaves the
 * user unable to tell whether to wait, to switch, or to configure it by hand. An unavailable
 * state that cannot name its cause and its way out is not worth rendering.
 *
 * It runs one check that `resolve` does not: the recipe has to be written for the exact
 * transport this model actually goes out on. A recipe authored for another protocol is the
 * same as no recipe at all.
 */
enum class CapabilityControlPresentation {
    AutomaticAvailable,
    ForceUnsupported,
    CustomOnly,
    Pending,

    /**
     * The provider really does have this capability, but only as a standalone external
     * service or MCP connector - it is not reachable from this connection's chat request
     * path. This has to stay separable from [Unsupported] (it simply does not exist):
     * folding both into one "not supported" hides the boundary and makes it look like a
     * shortcoming of the model itself.
     */
    ExternalConnectorOnly,
    Unsupported,
    Unknown;

    /**
     * Whether the user can express an intent here at all. [ForceUnsupported] still counts as
     * configurable: it only says the "force" position is unavailable, while off and automatic
     * remain requests we can genuinely send.
     */
    val isConfigurable: Boolean
        get() = when (this) {
            AutomaticAvailable, ForceUnsupported, Unknown -> true
            CustomOnly, Pending, ExternalConnectorOnly, Unsupported -> false
        }

    companion object {
        fun status(
            state: String?,
            reasonCode: String?,
            exactTransportMatches: Boolean,
            forceRequested: Boolean = false,
        ): CapabilityControlPresentation = when (state) {
            "auto_available" ->
                if (!exactTransportMatches) Unknown
                else if (forceRequested) ForceUnsupported
                else AutomaticAvailable
            "custom_only" -> CustomOnly
            "unavailable" ->
                if (reasonCode == "external_connector_only") ExternalConnectorOnly else Unsupported
            // `unknown` stays configurable; the reason code only refines the cause of that one
            // state and must never be allowed to promote an unknown into an unavailable.
            // Kill-switch and pending routes are read-only sub-states of unknown.
            "unknown", null -> when (reasonCode) {
                "endpoint_route_pending", "model_route_pending", "official_source_insufficient",
                "source_review_expired", "provider_kill_switch",
                -> Pending
                else -> Unknown
            }
            else -> Unknown
        }
    }
}

/**
 * The panel, the composer summary and the outbound gate all have to answer the same question:
 * does this capability have an automatic configuration right now? Three separate copies of
 * that judgement would start disagreeing sooner or later, so it lives here once.
 */
object CapabilityControlPresentationResolver {
    fun presentation(
        provider: Provider,
        model: AIModel,
        capability: String,
        metadata: MetadataClient = MetadataClient.instance,
        effectiveTransport: String? = null,
    ): CapabilityControlPresentation {
        val verdict = CapabilityControlResolution.resolve(provider, model, capability, metadata, effectiveTransport)
        return CapabilityControlPresentation.status(
            state = verdict.state,
            reasonCode = verdict.reasonCode,
            // A subscription link has no recipe, so the "recipe transport == catalog transport"
            // check has nothing to compare. Its verdict comes from what the upstream declares
            // for this specific model, and the protocol is pinned by the link itself (see
            // `CapabilityControlResolution.subscriptionFinalTransport`), so a mismatch is
            // structurally impossible. Without this exemption `status(...)` would downgrade
            // every `auto_available` back to Unknown and the verdict above would be wasted.
            exactTransportMatches = verdict.viaLegacyProfile ||
                CapabilityControlResolution.isSubscriptionLink(provider) ||
                hasExactTransportRecipe(provider, model, capability, metadata),
        )
    }

    /**
     * Whether the recipe's declared protocol and the protocol the catalog gives this model are
     * the same one. States other than `auto_available` have no recipe by contract, so this
     * check must not knock them down to unknown.
     */
    fun hasExactTransportRecipe(
        provider: Provider,
        model: AIModel,
        capability: String,
        metadata: MetadataClient = MetadataClient.instance,
    ): Boolean {
        val lookup = metadata.capabilityActionLookup(provider.kind, model.id, capability) ?: return false
        if (lookup.state != "auto_available") return true
        val recipeTransport = lookup.recipeTransport ?: return false
        val modelTransport = lookup.modelTransport ?: return false
        return canonicalCapabilityTransport(recipeTransport) == canonicalCapabilityTransport(modelTransport)
    }
}

/**
 * Stale-preference guard: will the stored web-search preference actually reach the wire right
 * now?
 *
 * The preference is stored per connection, model and transport, but whether it can go out
 * depends on the catalog metadata as it stands today. After the metadata rolls forward, a
 * recipe disappears, or the model moves to another transport, that stored `Automatic` has not
 * changed by a single byte - so the panel switch and the composer globe stay lit while the
 * request carries no web-search field at all. That is exactly how you end up with a lit globe
 * that never searches.
 *
 * This does not rewrite storage: the user's expressed intent stays, and comes back as soon as
 * they return to a model that supports it. It only stops lighting up a globe that cannot fire.
 */
object CapabilityWebPreferenceLiveness {
    /**
     * The pure predicate. Note it is deliberately *not* the same set as
     * [CapabilityControlPresentation.isConfigurable]: `ForceUnsupported` is still configurable
     * (off and automatic go out fine) but it does not light the globe right now.
     */
    fun reachesTheWire(
        status: CapabilityControlPresentation,
        customIsActive: Boolean,
    ): Boolean = status == CapabilityControlPresentation.AutomaticAvailable || customIsActive

    /**
     * Call sites outside the panel (the composer, the outbound gate) have no status at hand,
     * so this recomputes one from the same production source.
     *
     * @param forwardPortsStaleCustom also run the lazy forward-port of stale custom fragments.
     *   **Only discrete events may pass true** - the composer switching model, switching
     *   conversation, or closing the panel. On the recomposition hot path, ask the question but
     *   read only the conclusion: decoding SharedPreferences again on every recomposition is
     *   the classic "side effect inside the frame loop" mistake. The forward-port is idempotent,
     *   so firing it once per discrete event plus once on the send path is enough.
     */
    fun reachesTheWire(
        provider: Provider,
        model: AIModel,
        conversationID: String?,
        customFragmentStore: LocalCapabilityCustomFragmentStore,
        metadata: MetadataClient = MetadataClient.instance,
        effectiveTransport: String? = null,
        activeProfile: GenerationProfileRef? = null,
        forwardPortsStaleCustom: Boolean = false,
    ): Boolean {
        val status = CapabilityControlPresentationResolver.presentation(
            provider, model, "web", metadata, effectiveTransport,
        )
        if (status == CapabilityControlPresentation.AutomaticAvailable) return true
        val identity = ModelControlRuntimeIdentityResolver.resolve(provider, model, metadata) ?: return false
        // "Has a custom fragment taken over web right now" must mean exactly the same thing
        // here, in the UI and on the outbound path: enabled **and** this connection, model and
        // transport still has a usable web schema. Once the schema is withdrawn that JSON no
        // longer goes out (`fragmentsByOwner` already filters on the same predicate), so a lit
        // globe would be one more piece of fake state.
        val customIsActive = capabilityCustomFragmentAvailable(
            providerKind = provider.kind,
            modelID = model.id,
            finalTransport = identity.finalTransport,
            activeProfile = activeProfile,
            owner = "web",
        ) && customFragmentStore.effectiveConfiguration(
            providerID = provider.id,
            modelID = identity.canonicalModelId,
            conversationID = conversationID,
            transportIdentity = identity.storageIdentity,
            namespace = LocalCapabilityCustomFragmentStore.WEB_NAMESPACE,
            // Whether the globe lights depends on whether the fragment stored under an older
            // recipe version still counts.
            forwardPort = if (forwardPortsStaleCustom) {
                LocalCapabilityCustomFragmentStore.ForwardPortContext(
                    providerKind = provider.kind,
                    schemaModelID = model.id,
                    activeProfile = activeProfile,
                )
            } else {
                null
            },
        ).enabled
        return reachesTheWire(status, customIsActive)
    }
}

File: web/apps/app/lib/core/chat/capability-control-presentation.ts
import type { AIModel, Provider } from '@oriveo/shared';
import { resolveControl } from '@oriveo/core/providers/request-preference/capability-runtime';
import { canonicalRecipeTransport } from '@oriveo/core/providers/request-builders/capability-execution';
import type { CapabilityKey } from '@oriveo/core/providers/request-preference/types';
import { codexReasoningEffort } from '@oriveo/core/providers/openai-subscription';
import {
  getCapabilityRuntime,
  getDeclaredReasoningLevels,
  getModelTransport,
  subscriptionDeclaredReasoningLevels,
} from '../metadata/metadata-client';
import { currentCapabilityEvidenceModel, resolveModelCapabilityEvidence } from './capability-evidence';
import { customFragmentOwnerFacts } from './custom-fragment-settings';
import { modelControlWebReachesTheWire, resolveModelControlStatus } from './model-control-capability-layout';
import { subscriptionFinalTransport } from './capability-preference-settings';

export type CapabilityControlPresentation = {
  state: 'auto_available' | 'managed_only' | 'custom_only' | 'unavailable' | 'unknown';
  availableIntents: readonly string[];
  /** Stable runtime verdict, for localized UI diagnostics only (never raw server prose). */
  reasonCode?: string;
  /**
   * true when this verdict came from a legacy profile fallback rather than a v2 recipe, so
   * diagnostics can tell them apart. Comparing `reasonCode === 'legacy_profile'` would be
   * equivalent but more brittle: renaming the constant silently loses the signal and the type
   * system cannot catch it. reasonCode stays reserved for user-facing text.
   */
  viaLegacyProfile: boolean;
};

/** Only local message keys cross the UI boundary; Server reasonCode is never prose. */
export function capabilityControlReasonMessageKey(
  state: CapabilityControlPresentation['state'],
  reasonCode: string | undefined,
):
  | 'capabilityControlReasonPending'
  | 'capabilityControlReasonUnsupported'
  | 'capabilityControlReasonExternalConnector'
  | 'capabilityControlReasonCustom'
  | 'capabilityControlReasonUnknown' {
  // State is the Server verdict. reasonCode is an open, supplementary diagnostic
  // vocabulary and must never turn custom_only into unavailable (or vice versa).
  if (state === 'custom_only') return 'capabilityControlReasonCustom';
  if (state === 'unknown') return 'capabilityControlReasonPending';
  if (reasonCode === 'relay_user_directory') return 'capabilityControlReasonCustom';
  if (reasonCode === 'endpoint_route_pending' || reasonCode === 'model_route_pending'
    || reasonCode === 'official_source_insufficient' || reasonCode === 'source_review_expired'
    || reasonCode === 'provider_kill_switch') return 'capabilityControlReasonPending';
  // The provider does have this capability, but only as a separate external service / MCP
  // connector outside this connection's chat request. Folding it into the generic
  // "not supported" copy would hide the boundary from the user.
  if (reasonCode === 'external_connector_only') return 'capabilityControlReasonExternalConnector';
  if (reasonCode === 'no_official_managed_search' || reasonCode === 'transport_not_supported'
    || reasonCode === 'model_capability_absent' || reasonCode === 'retired_model_alias'
    || reasonCode === 'upstream_parameter_not_declared') return 'capabilityControlReasonUnsupported';
  return state === 'unavailable'
    ? 'capabilityControlReasonUnsupported'
    : 'capabilityControlReasonUnknown';
}

const LEGACY_REASONING_INTENTS: Record<string, string> = {
  fast: 'low', balanced: 'balanced', deep: 'deep', max: 'max',
};

/**
 * The legacy request profile from bundled provider config `.profiles`.
 *
 * This is NOT a client-side model-id heuristic — it is the same Server-authored
 * profile the request builders actually dispatch (`resolveWebSearchProfile` /
 * `resolveReasoningParams`). Reading it back is therefore allowed where the rule against
 * guessing applies: what is banned is inventing a capability from a model id, not consuming a
 * capability the catalog already published.
 */
/** Kept for readers that still consume the older profile shape; the current presentation does not. */
export function legacyCapabilityFallback(
  provider: Provider,
  model: AIModel,
  capability: CapabilityKey,
  resolvedModel?: ReturnType<typeof currentCapabilityEvidenceModel>,
): CapabilityControlPresentation | undefined {
  // Relay directories come from the user's own machine; they never carry an
  // official Server profile, so there is nothing authoritative to fall back to.
  if (provider.kind === 'relay') return undefined;
  const current = resolvedModel ?? currentCapabilityEvidenceModel(provider, model);
  const evidenceSupported = (key: 'web_search' | `reasoning_level/${string}`) => (
    resolveModelCapabilityEvidence({ key, provider, model }).support === 'supported'
  );
  if (capability === 'web') {
    // Manually added models and older persisted records can be missing the capabilities array entirely, so it must not be assumed present.
    const byProfile = (current.capabilities ?? []).includes('web') && Boolean(current.webSearchProfile);
    return byProfile || evidenceSupported('web_search')
      ? { state: 'auto_available', availableIntents: ['off', 'automatic'], reasonCode: 'legacy_profile', viaLegacyProfile: true }
      : undefined;
  }
  if (capability === 'reasoning') {
    // Declared profile levels first; the evidence facade is the second source so
    // a future Server-published `reasoning_level/*` verdict still lights the control
    // up without a profile. Today's production metadata publishes neither
    // `reasoning_level/*` nor `web_search` candidates, so profiles carry it alone.
    const levels = getDeclaredReasoningLevels(current.reasoningProfile);
    const byEvidence = levels.length === 0
      ? (Object.keys(LEGACY_REASONING_INTENTS) as (keyof typeof LEGACY_REASONING_INTENTS)[])
        .filter((level) => evidenceSupported(`reasoning_level/${level}`))
      : [];
    const effective = levels.length > 0 ? levels : byEvidence;
    return effective.length > 0
      ? {
          state: 'auto_available',
          availableIntents: effective.map((level) => LEGACY_REASONING_INTENTS[level] ?? level),
          reasonCode: 'legacy_profile',
          viaLegacyProfile: true,
        }
      : undefined;
  }
  return undefined;
}

/**
 * Capability verdict for the subscription paths (Grok subscription and Codex).
 *
 * Subscription models are not in the metadata catalog and never have a v2 control. Falling back to
 * `unknown` would make the UI render "not adjustable" for a capability the upstream explicitly
 * declares, leaving the user with no switch at all. The authority for these paths is not the
 * catalog anyway: the subscription's own `/models` endpoint declares capabilities per model, and
 * that is the only legitimate authority here.
 *
 * Both subscription paths allow a capability only when the upstream declares it per model. Grok
 * additionally requires the final protocol to be Responses; an explicit chat declaration has no
 * real web_search and must stay unavailable.
 */
function subscriptionCapabilityControl(
  provider: Provider,
  model: AIModel,
  capability: CapabilityKey,
): CapabilityControlPresentation {
  if (capability === 'web') {
    if (provider.kind === 'grok' && subscriptionFinalTransport(provider, model) !== 'openai_responses') {
      return {
        state: 'unavailable', availableIntents: [],
        reasonCode: 'upstream_transport_without_web_search', viaLegacyProfile: false,
      };
    }
    return (model.capabilities ?? []).includes('web')
      ? {
          state: 'auto_available',
          availableIntents: provider.kind === 'grok' ? ['automatic'] : ['off', 'automatic'],
          reasonCode: 'subscription_upstream_declared', viaLegacyProfile: false,
        }
      : {
          state: 'unavailable', availableIntents: [],
          reasonCode: 'model_capability_absent', viaLegacyProfile: false,
        };
  }
  if (capability === 'reasoning') {
    // Filter the available levels with the same function used on the outbound path, so every level
    // shown in the UI maps to a value the upstream actually accepts. A second mapping would drift,
    // which is exactly how the UI ends up offering a choice that is sent as something else.
    const declared = subscriptionDeclaredReasoningLevels(provider.kind, model);
    const intents = (Object.keys(LEGACY_REASONING_INTENTS) as Array<keyof typeof LEGACY_REASONING_INTENTS>)
      .filter((mode) => codexReasoningEffort(mode, declared) !== undefined)
      .map((mode) => LEGACY_REASONING_INTENTS[mode] ?? mode);
    return intents.length > 0
      ? {
          state: 'auto_available', availableIntents: intents,
          reasonCode: 'subscription_upstream_declared', viaLegacyProfile: false,
        }
      : {
          state: 'unavailable', availableIntents: [],
          reasonCode: 'upstream_parameter_not_declared', viaLegacyProfile: false,
        };
  }
  // Every other capability has no declared source on the subscription paths, so report unknown honestly.
  return { state: 'unknown', availableIntents: [], viaLegacyProfile: false };
}

/**
 * The single capability verdict for one provider/model/capability. Both the chat
 * composer controls and the model-list badges resolve through this function, so
 * "the library says it supports Web / the composer cannot turn Web on" is
 * structurally impossible rather than merely tested against.
 *
 * Precedence:
 *   1. A valid v2 control decides, including an explicit `unavailable` — a
 *      recipe that positively says "this transport/model cannot do it" outranks
 *      any older profile.
 *   2. `unknown` or a missing runtime remains honest `unknown`; new clients do
 *      not revive legacy profiles. The reader above exists only for older callers.
 */
export function presentCapabilityControl(
  provider: Provider | undefined,
  model: AIModel | undefined,
  capability: CapabilityKey,
  /** Pass an already-resolved catalog projection to avoid re-resolving it per capability. */
  resolvedModel?: ReturnType<typeof currentCapabilityEvidenceModel>,
): CapabilityControlPresentation {
  void resolvedModel;
  if (!provider || !model) return { state: 'unknown', availableIntents: [], viaLegacyProfile: false };
  // Subscription paths exit early: their capability authority is the upstream /models declaration,
  // not the catalog (see the function comment above). This must come before the catalog transport
  // check, because subscription models are not in the catalog at all and would only fall to unknown.
  if (provider.authMode === 'subscription') {
    return subscriptionCapabilityControl(provider, model, capability);
  }
  // A persisted model can survive a metadata refresh. Do not present a
  // recipe when the selected model and catalog would dispatch different transports.
  const catalogTransport = getModelTransport(model.id, provider.kind);
  if (catalogTransport && model.transport && catalogTransport !== model.transport) {
    return { state: 'unknown', availableIntents: [], reasonCode: 'transport_mismatch', viaLegacyProfile: false };
  }
  const runtime = getCapabilityRuntime();
  const control = model.capabilityControls?.[capability];
  if (!runtime || !control) {
    return { state: 'unknown', availableIntents: [], viaLegacyProfile: false };
  }
  const resolved = resolveControl(provider.kind, capability, control, {
    recipes: Object.keys(runtime.recipes ?? {}),
    sourceIndex: runtime.sourceIndex ?? {},
    controlDefinitions: runtime.controlDefinitions ?? {},
  });
  const state = ['auto_available', 'managed_only', 'custom_only', 'unavailable', 'unknown'].includes(resolved.state)
    ? resolved.state as CapabilityControlPresentation['state']
    : 'unknown';
  return {
    state,
    availableIntents: resolved.valid && Array.isArray(control.availableIntents)
      ? control.availableIntents
      : [],
    ...(resolved.reason ?? control.reasonCode ? { reasonCode: resolved.reason ?? control.reasonCode } : {}),
    viaLegacyProfile: false,
  };
}

/**
 * Call sites outside the panel (chip highlighting, restore state, the explicit outbound intent key)
 * have no ready status or custom facts, so recompute once from the production source. The rule
 * itself is only `modelControlWebReachesTheWire`; this just gathers the facts.
 *
 * This is meant for discrete events (effects, callbacks, the send path), not the render hot path:
 * it parses a metadata snapshot and reads the local custom configuration once.
 */
export function webPreferenceReachesTheWire(input: {
  provider?: Provider;
  model?: AIModel;
  transportIdentity?: string;
}): boolean {
  const status = resolveModelControlStatus(presentCapabilityControl(input.provider, input.model, 'web'));
  if (status === 'automaticAvailable') return true;
  if (!input.provider || !input.model || !input.transportIdentity) return false;
  return modelControlWebReachesTheWire({
    status,
    customIsActive: customFragmentOwnerFacts({
      provider: input.provider, model: input.model, transportIdentity: input.transportIdentity, owner: 'web',
    }).isActive,
  });
}

/**
 * Whether the protocol declared by this model recipe is the same one the catalog gives it.
 *
 * `presentCapabilityControl` answers "what is the server verdict for this capability", which is one
 * check short of this one: it does not verify the exact transport - **a recipe written for another
 * protocol is as good as absent**. At outbound compile time `resolveRuntimeRecipe` drops it with
 * `transport_mismatch`, so switching to that model would only lead to a second dead end. The
 * candidate rule behind "view supported models" therefore has to ask this extra question.
 *
 * States other than `auto_available` have no recipe by contract, so this check must not mark them unavailable.
 */
export function hasExactCapabilityTransportRecipe(
  provider: Provider | undefined,
  model: AIModel | undefined,
  capability: CapabilityKey,
): boolean {
  if (!provider || !model) return false;
  const control = model.capabilityControls?.[capability];
  if (!control) return false;
  if (control.state !== 'auto_available') return true;
  const runtime = getCapabilityRuntime();
  const recipeRef = control.recipeRef;
  const modelTransport = getModelTransport(model.id, provider.kind) ?? model.transport;
  if (!runtime || !recipeRef || !modelTransport) return false;
  const recipe = runtime.recipes?.[recipeRef];
  if (!recipe || typeof recipe !== 'object' || Array.isArray(recipe)) return false;
  const transport = (recipe as { transport?: { protocol?: string } }).transport?.protocol;
  return canonicalRecipeTransport(transport) === canonicalRecipeTransport(modelTransport);
}

/** D0: missing capability knowledge is still user-configurable; only a negative verdict disables it. */
export function capabilityControlIsConfigurable(
  control: Pick<CapabilityControlPresentation, 'state'>,
): boolean {
  return control.state === 'auto_available' || control.state === 'unknown';
}

/**
 * The badge/filter verdict for model lists, resolved from the same function the
 * composer uses. `custom_only` deliberately does not earn a badge: the user
 * cannot turn it on without the developer switch.
 */
export function capabilityAvailableForDisplay(
  provider: Provider | undefined,
  model: AIModel | undefined,
  capability: CapabilityKey,
  resolvedModel?: ReturnType<typeof currentCapabilityEvidenceModel>,
): boolean {
  const state = presentCapabilityControl(provider, model, capability, resolvedModel).state;
  return state === 'auto_available' || state === 'managed_only';
}

/**
 * Available web search levels are taken from `availableIntents` only under an official
 * `auto_available` recipe; every other state yields an empty list, because a `force` level without
 * an official declaration should not appear. This is deliberately asymmetric with how reasoning is
 * read (reasoning uses the resolve result directly) and must not be "unified". It lives in this
 * module because v2 state words are only allowed inside the registered verdict surface; rendering
 * components must not carry bare checks of their own.
 */
export function resolveWebAvailableIntents(control: CapabilityControlPresentation): readonly string[] {
  return control.state === 'auto_available' ? control.availableIntents : [];
}
